### PR TITLE
Dashroll implementation

### PR DESCRIPTION
**Pull request by RyRy79261**
# Summary

Dash implementation was achieved with percentage modifiers to allow for easing functions later on. Cool down was achieved by retaining the coroutine's loop for a period after the movement effects.

## Possible improvements

- Direction checks in the coroutine could be improved to be less redundant with movement or potentially have a movement script which applies the effect in a better way 
- Dash triggering via PlayerActions, I wasn't sure if, via this input triggering, additional attempts to execute the dash were being blocked.
- Left & Right direction could have slightly shorter dash times

### CODE REVIEW

File: Assets/Scripts/Player/PlayerController.cs [hunk 1]
```diff
@@ -301,4 +307,76 @@ private void Update()
...
+    IEnumerator DashRoll()
+    {
+        float dashAnimation = Time.time + dashDuration;
+        float dashComplete = dashAnimation + dashCooldown;
+        // Default is forward
+        DashDirection direction = DashDirection.Forward;
+
+        float dashedSpeed = moveSpeed * dashSpeedModifier;
+
+        if (Input.GetKey(KeyCode.S))
```
[the remark this answers is not in the document]
Comment (RyRy79261) [reply Assets/Scripts/Player/PlayerController.cs:330]: @kantagara Yeah our system looks extremely different to the tutorials I was following. Is the new system added yet?

[the remark this answers is not in the document]
Comment (sneakzttv) [reply Assets/Scripts/Player/PlayerController.cs:330]: not yet, this is fine tbh i can alter the input later.

**Review by sneakzttv — APPROVED** [review]

File: Assets/Scripts/Player/PlayerController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sneakzttv) [reply Assets/Scripts/Player/PlayerController.cs:330]: image

[the remark this answers is not in the document]
Comment (RyRy79261) [reply Assets/Scripts/Player/PlayerController.cs:330]: Like preferably, isDodgeRolling would be an enum of either left, right, back, forward, idle. That's set in the event triggered by the dashbutton, and in said event determines if theres a specific direction other than default forward and sets it

[the remark this answers is not in the document]
Comment (sneakzttv) [reply Assets/Scripts/Player/PlayerController.cs:330]: fixed will leave reviews :D

File: Assets/Scripts/Player/PlayerController.cs [hunk 2]
```diff
@@ -301,4 +307,76 @@ private void Update()
...
+                    {
+                        controller.Move(transform.forward * (dashedSpeed * Time.deltaTime));
+                        break;
+                    }
+                }
+            }
+            yield return null;
+        }
+
+        isDodgeRolling = false;
```
Comment (RyRy79261) [inline Assets/Scripts/Player/PlayerController.cs:377]:
```suggestion
        isDodgeRolling = false;
        // Grapes 
```

File: Assets/Scripts/Player/PlayerController.cs [hunk 3]
```diff
@@ -195,19 +200,19 @@ private void Move()
     {
```
Comment (sneakzttv) [inline Assets/Scripts/Player/PlayerController.cs:200]:
```suggestion
    /// <summary>
    /// Handle player movement
    /// </summary>
    private void Move()
    {
        controller.Move(transform.rotation * currentMovement * moveSpeed * Time.deltaTime);
        
        if (isMovementPressed)
        {
            // animations for later go here following this syntax
            //animator.SetBool("isRunning", true);
        }
        else
        {
            //animator.SetBool("isRunning", false);
        }
        velocity.y += gravity * Time.deltaTime;
```

File: Assets/Scripts/Player/PlayerController.cs [hunk 4]
```diff
@@ -301,4 +307,76 @@ private void Update()
     }
     
     #endregion
+
+    // Consider moving all Behaviours into seperate script
```
Comment (sneakzttv) [inline Assets/Scripts/Player/PlayerController.cs:311]:
```suggestion
   #region Dash

IEnumerator DashRoll()
{
    float dashAnimation = Time.time + dashDuration;
    float dashComplete = dashAnimation + dashCooldown;
    float dashedSpeed = moveSpeed * dashSpeedModifier;
    
    while (Time.time < dashComplete)
    {
        // This includes the cool down in the while loop, but only updates the movement during the intended animation sequence 
        if (Time.time < dashAnimation)
        {
            controller.Move(transform.rotation * currentMovement * (dashedSpeed * Time.deltaTime));
        }
        yield return null;
    }

    isDodgeRolling = false;
}

#endregion
```

Comment (RyRy79261) [reply Assets/Scripts/Player/PlayerController.cs:311]: Lol, `transform.rotation` I feel like I'm in CSS again 😂 

Thanks!

**Review by kantagara — CHANGES_REQUESTED** [review]
There are couple of things that should be sorted out:

1. Increase Dash Speed Modifier to something like 5 or 6
2. Remove the boolean isDodgeRolling, introduce a new field of type Coroutine called dashRollCoroutine.  Because currently the coroutines were being called repeatedly, and that shouldn't be an indented behavior.
3. ```/// <summary>
    /// Used for player dodge rolling input
    /// </summary>
    private void OnDodgeRollInput(InputAction.CallbackContext context)
    {
        if (dashRollCoroutine == null)
            dashRollCoroutine = StartCoroutine(DashRoll());
    }``` the OnDodgeRollInput method should look like this
4. Make a unified naming, is it dodge roll or dash roll (I would say Dash Roll)
5. ```IEnumerator DashRoll()
    {
        float dashAnimation = Time.time + dashDuration;
        float dashComplete = dashAnimation + dashCooldown;
        float dashedSpeed = moveSpeed * dashSpeedModifier;
    
        while (Time.time < dashComplete)
        {
            // This includes the cool down in the while loop, but only updates the movement during the intended animation sequence 
            if (Time.time < dashAnimation)
            {
                controller.Move(transform.rotation * currentMovement * (dashedSpeed * Time.deltaTime));
            }
            yield return null;
        }
        dashRollCoroutine = null;
    }``` dashRoll Coroutine should look like this. 
6. Remove DodgeRoll method from Update();

Later on we'll most likely  move towards the state machine implementation, but for now I think this would be enough.

**Review by sneakzttv — APPROVED** [review]

**Pull request merged**